### PR TITLE
hostapd: Add support for APuP

### PR DESCRIPTION
**Pull request by G10h4ck**
Add support for hostapd Access Point Micro Peering

### CODE REVIEW

**Comment by tapper82** [comment]
Access Point Micro Peering is a simpler and hopefully more useful successor to
Ad Hoc, Wireless Distribution System, 802.11s mesh mode, Multi-AP and EasyMesh.
When enabled almost plain APs communicate between them via 4-address mode,
like in WDS but all of them are AP, so they can eventually communicate also with
plain stations and more AP nodes in sight, without more trickery.
APuP has low hardware requirements, just AP mode support + 4-address mode, and
no more unnecessary complications, like hardcoded bridging or routing algorithm
in WiFi stack.
For each AP in sight an interface is created, and then it can be used as
convenient in each case, bridging, routing etc.
Those interfaces could be simply bridged in a trivial topology (which happens
automatically if wds_bridge is not an empty string), or feeded to a
routing daemon.

**Comment by hauke** [comment]
How stable is it working? 
I have heard of some problems when an access point disconnects that it is not fully removed for example.

**Comment by G10h4ck** [comment]
> How stable is it working? I have heard of some problems when an access point disconnects that it is not fully removed for example.

I attempted to reproduce the reported issue by @mwarning but couldn't reproduce it, and after some months even @mwarning wrote to me that he is not able to reproduce that issue anymore, so most probably the issue that was reported was due to an unrelated problem in the driver that have been fixed since then

**Comment by hauke** [comment]
Thanks for the clarification.

The github CI found some compile problems:
```
====== Make errors from logs/package/network/services/hostapd/wpad-mesh-openssl/compile.txt ======
  CC  ucode.c
  CC  ../src/rsn_supp/wpa_ft.c
  CC  mesh.c
mesh_mpm.c:122:12: error: conflicting types for 'copy_supp_rates'; have 'u16(struct wpa_supplicant *, struct sta_info *, struct ieee802_11_elems *)' {aka 'short unsigned int(struct wpa_supplicant *, struct sta_info *, struct ieee802_11_elems *)'}
  122 | static u16 copy_supp_rates(struct wpa_supplicant *wpa_s,
      |            ^~~~~~~~~~~~~~~
In file included from mesh_mpm.c:18:
../src/ap/ieee802_11.h:111:5: note: previous declaration of 'copy_supp_rates' with type 'u16(struct hostapd_data *, struct sta_info *, const struct ieee802_11_elems *)' {aka 'short unsigned int(struct hostapd_data *, struct sta_info *, const struct ieee802_11_elems *)'}
  111 | u16 copy_supp_rates(struct hostapd_data *hapd, struct sta_info *sta,
      |     ^~~~~~~~~~~~~~~
make[4]: *** [../src/build.rules:89: /__w/openwrt/openwrt/openwrt/build_dir/target-mips-openwrt-linux-musl_musl/hostapd-wpad-mesh-openssl/hostapd-2024.03.09~695277a5/build/wpa_supplicant/mesh_mpm.o] Error 1
make[4]: *** Waiting for unfinished jobs....
  CC  mesh_rsn.c
make[4]: Leaving directory '/__w/openwrt/openwrt/openwrt/build_dir/target-mips-openwrt-linux-musl_musl/hostapd-wpad-mesh-openssl/hostapd-2024.03.09~695277a5/wpa_supplicant'
make[3]: *** [Makefile:817: /__w/openwrt/openwrt/openwrt/build_dir/target-mips-openwrt-linux-musl_musl/hostapd-wpad-mesh-openssl/hostapd-2024.03.09~695277a5/.built] Error 2
time: package/network/services/hostapd/wpad-mesh-openssl/compile#21.27#7.47#15.49
```

Have you tried to get this into upstream hostapd?

File: package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+ 
+ #ifdef CONFIG_FILS
+ static struct wpabuf *
+@@ -3469,8 +3472,8 @@ static u16 check_multi_ap(struct hostapd
+ }
+ 
+ 
+-static u16 copy_supp_rates(struct hostapd_data *hapd, struct sta_info *sta,
+-			   struct ieee802_11_elems *elems)
++u16 copy_supp_rates(struct hostapd_data *hapd, struct sta_info *sta,
```
Comment (mwarning) [inline package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch:360]: @G10h4ck this function conflicts with a different function of the same name on some targets. A simple solution would be to expose it with a different name.

**Comment by mwarning** [comment]
@hauke The crash appeared when a mesh interface got removed (timeout). But since that is not happening anymore, this feature looks save to work with. Mayb it was only an incompatible hostapd that I used as base.

Anyway, this is a nice feature and inclusion in to OpenWrt will give this very promising feature more exposure.
A try to push this upstream would also allow more knowledgeable people to have a look.

**Comment by G10h4ck** [comment]
> Have you tried to get this into upstream hostapd?

Yes, but the only comments I received was about space/tab indentation

File: package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch (hunk 1 above)
Comment (G10h4ck) [reply package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch:360]: yeah, definitely this seems the easiest fix, but I am not sure if the right thing would be to rename this function or all the conflicting others...

Comment (mwarning) [reply package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch:360]: You could introduce a new function that calls the private copy_supp_rates. That would minimize the changes.

**Comment by hauke** [comment]
I think the diff is getting too large. Could you try to rebase on current master.

**Comment by mwarning** [comment]
The pipeline outputs "get_changed_packages.pl: Argument list too long". hmm

**Comment by mwarning** [comment]
> I think the diff is getting too large. Could you try to rebase on current master.

Do you mean that the amount of changes is too large?

**Comment by G10h4ck** [comment]
> I think the diff is getting too large. Could you try to rebase on current master.

done, let's see what come out from the CI this time ;-)

**Comment by G10h4ck** [comment]
@tapper82  @hauke @mwarning should be fine now as you can see on the CI run on my fork https://github.com/G10h4ck/openwrt/tree/apup

File: package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+ 		return -1;
++
++	const char *bridge = NULL;
++	char ifName[IFNAMSIZ + 1];
++
++	int mRet = 0;
++
++#ifdef CONFIG_APUP
++	if(hapd->conf->apup && hapd->conf->apup_peer_ifname_prefix[0])
++	{
```
Comment (mwarning) [inline package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch:132]: Some nitpicking. Let's use the same coding style as the rest of the code.

`if (hapd->conf->apup && hapd->conf->apup_peer_ifname_prefix[0]) {`

File: package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
++		            hapd->conf->iface, aid );
++
++	if (mRet >= (int) sizeof(ifName))
++		wpa_printf( MSG_WARNING,
++		            "nl80211: WDS interface name was truncated" );
++	else if (mRet < 0)
++		return mRet;
++
++	// Pass back to the caller the resulting interface name
++	if (ifname_wds) os_strlcpy(ifname_wds, ifName, IFNAMSIZ + 1);
```
Comment (mwarning) [inline package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch:150]: os_strlcpy should go on a separate line

File: package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
++
+ 	if (hapd->conf->wds_bridge[0])
+ 		bridge = hapd->conf->wds_bridge;
++
```
Comment (mwarning) [inline package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch:154]: Let's not try to touch lines outside of the patched lines. This will make it less likely that the patch will conflict with other changes in the future.

File: package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
+index aeb1e6beef..bd9a9b3b7e 100644
+--- a/src/drivers/driver_nl80211.c
++++ b/src/drivers/driver_nl80211.c
+@@ -8415,24 +8415,14 @@ static int have_ifidx(struct wpa_driver_nl80211_data *drv, int ifidx,
+ 
+ 
+ static int i802_set_wds_sta(void *priv, const u8 *addr, int aid, int val,
+-			    const char *bridge_ifname, char *ifname_wds)
++                const char *bridge_ifname, const char *ifname_wds)
```
Comment (mwarning) [inline package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch:439]: This indentation change is not needed.

File: package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch [hunk 6]
```diff
@@ -0,0 +1,463 @@
...
+ int hostapd_set_wds_sta(struct hostapd_data *hapd, char *ifname_wds,
+ 			const u8 *addr, int aid, int val)
+ {
+-	const char *bridge = NULL;
+-
+ 	if (hapd->driver == NULL || hapd->driver->set_wds_sta == NULL)
+ 		return -1;
++
++	const char *bridge = NULL;
++	char ifName[IFNAMSIZ + 1];
```
Comment (hauke) [inline package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch:126]: Please keep the variable definition at the beginning of the function.

File: package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch (hunk 3 above)
Comment (hauke) [inline package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch:150]: Why did you move the code to this location?

**Comment by tapper82** [comment]
When this is ready, can we get some one to do a write up in the wiki? Lets not have this a feature that people don't use because they don't know about it.

File: package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch (hunk 6 above)
Comment (G10h4ck) [reply package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch:126]: done

File: package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch (hunk 2 above)
Comment (G10h4ck) [reply package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch:132]: done

File: package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch (hunk 3 above)
Comment (G10h4ck) [reply package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch:150]: done

Comment (G10h4ck) [reply package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch:150]: Because in the original hostapd code the WDS sta iface name is hardcoded in the driver which is not reasonable in general, and specifically problematic for APuP use case, so with @nbd168 we decided to make sta interface name configurable with a sane default. And to inform back what is the interface name, when it is created. So it is a needed change.

File: package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch (hunk 4 above)
Comment (G10h4ck) [reply package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch:154]: same reason as previous note

Because in the original hostapd code the WDS sta iface name is hardcoded in the driver which is not reasonable in general, and specifically problematic for APuP use case, so with @nbd168 we decided to make sta interface name configurable with a sane default. And to inform back what is the interface name, when it is created. So it is a needed change.

File: package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch (hunk 5 above)
Comment (G10h4ck) [reply package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch:439]: pfff damn hostapd code convention, this is not really indentation but aligment, so moderns editor change it automatically... let me fix it with a plain text editor...

Comment (G10h4ck) [reply package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch:439]: done

**Comment by G10h4ck** [comment]
@tapper82  @hauke @mwarning addressed the reviews comments

**Comment by G10h4ck** [comment]
> When this is ready, can we get some one to do a write up in the wiki? Lets not have this a feature that people don't use because they don't know about it.

IMHO this feature will even give competitive advantage to OpenWrt over other wifi router, so no worry it doesn't get used ;-)

hopefully at some point it would get enough traction to be included into the WiFi standard, as it should have be since decades...

**Comment by ascendbeing** [comment]
> When this is ready, can we get some one to do a write up in the wiki? Lets not have this a feature that people don't use because they don't know about it.

That's a good idea but also there's a lot of functionality that can be used that isn't easily accessible.

Biggest one is wpa_supplicant manual configuration. The block generated by Luci doesn't correspond to everything in full you plug in. You have to manually edit I think mac80211.sh or hostapd.sh to access quite a few fairly useful options (forcing 20mhz, turning off short gi client side only, etc).

And the same goes for hostapd. However it is WYSIWYG. It doesn't ignore anything you plug in. There are useful missing options. Probably the biggest is no way to use a space for individual options for hostapd_options. Some options in hostapd.conf, such as chanlist, want space separated channel list. You can't do that afaik. So if for example your AP uses lower dBm for certain channels theres no way to set chanlist to auto channel except those channels. But hostapd_options at least mostly works other then that. There's no wpa supplicant equivalent I'm aware of. unfortunately some options are always set after what you enter there meaning you'd have to manually edit if you need to have certain overrides be actually respected

File: package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+ 	if (resp != WLAN_STATUS_SUCCESS)
+ 		return resp;
+ 
+@@ -5927,6 +5930,10 @@ static void handle_beacon(struct hostapd_data *hapd,
+ 				      0);
+ 
+ 	ap_list_process_beacon(hapd->iface, mgmt, &elems, fi);
++
++#ifdef CONFIG_APUP
++	if(hapd->conf->apup) apup_process_beacon(hapd, mgmt, len, &elems);
```
Comment (mwarning) [inline package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch:396]: There are still a few places that have a different coding style compared to the rest of the code.
Please make sure that there is a space after the `if` and no space after the `(`.
@hauke I do not know how strict OpenWrt is regarding code style. But I think it makes it slightly more likely to be adapted upstream.

File: package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+diff --git a/src/drivers/driver.h b/src/drivers/driver.h
+index 1bbb9672cd..cbcb320b49 100644
+--- a/src/drivers/driver.h
++++ b/src/drivers/driver.h
+@@ -3976,7 +3976,7 @@ struct wpa_driver_ops {
+ 	 * Returns: 0 on success, -1 on failure
+ 	 */
+ 	int (*set_wds_sta)(void *priv, const u8 *addr, int aid, int val,
+-			   const char *bridge_ifname, char *ifname_wds);
++	           const char *bridge_ifname, const char *ifname_wds);
```
Comment (mwarning) [inline package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch:423]: @G10h4ck the semantics of ifname_wds are changed here. Before it was used as input and output. But now it is only used as input. Won't that break code in a hard to detect manner?

Comment (mwarning) [reply package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch:423]: hm, nevermind. The method seem to be called from one location only and you took care of that. So that's fine. :-)

File: package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch (hunk 4 above)
Comment (mwarning) [reply package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch:151]: Ah sorry,  I wasn't clear enough. I meant the newline that you added in line 151. It is outside the code that you added.

**Comment by mwarning** [comment]
> > When this is ready, can we get some one to do a write up in the wiki? Lets not have this a feature that people don't use because they don't know about it.
> 

Would you be willing to write some documentation for the wiki? That would be awesome. :-)

**Comment by ascendbeing** [comment]
> > When this is ready, can we get some one to do a write up in the wiki? Lets not have this a feature that people don't use because they don't know about it.
> 
> @ascendbeing Would you be willing to write some documentation for the wiki? That would be awesome. :-)

I'm more of an armchair complainer than a wiki author. I concede it would be useful to have a centralized spot for discussions or directions of where one might be able to put their efforts into, with regard to "advanced" (more usually, to try to stretch legacy/aging hardware to last a little longer into the future) configuration. So it would probably be useful to specifically lay out which options from hostapd.conf and wpa_supplicant.conf that could be useful. And it would probably be useful to show the current level of support or said options: support level a) supported by Luci; support level b) unsupported by Luci but supported by uci via hostapd_options; or support level c) unsupported by Luci and unsupported by uci hostapd_options but possible to implement by editing this or that file (hostapd.sh, mac80211.sh, etc.).
Probably the most near term thing to address my armchair complaints is to create a wpa_supplicant_options (whatever name is appropriate) equivalent of hostapd_options. Again I'm not an expert on the syntax of uci so I don't know if there's a way of inserting a space; hostapd_options seems to be "key=val key2=val2" where a space is meant to separate lines in the generated .conf file. So this is again just a near term thing that would help unlock more possibilities for people.
This is what I'm talking about btw: https://w1.fi/cgit/hostap/plain/hostapd/hostapd.conf and https://w1.fi/cgit/hostap/plain/wpa_supplicant/wpa_supplicant.conf and btw those files are of current HEAD so you have to be mindful of that owrt uses particular snapshots so you would have to go back a little bit. and there are several options owrt patches in like ht_coex which are documented on the wiki page about /etc/config/wireless

So that's a qualified maybe. Probably I'm not the right person to do it. I can't even format GitHub so I doubt I can format a wiki page, without learning up.

I know this is getting on a tangent (going off topic) so I'll leave it at that for now. I will check back if you have any thoughts on it though.

File: package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch (hunk 8 above)
Comment (G10h4ck) [reply package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch:423]: also`const` should take care of that too ;-)

File: package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch (hunk 7 above)
Comment (G10h4ck) [reply package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch]: should be done now

File: package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch [hunk 9]
```diff
@@ -0,0 +1,462 @@
...
+ 		bss->mld_indicate_disabled = atoi(pos);
+ #endif /* CONFIG_TESTING_OPTIONS */
+ #endif /* CONFIG_IEEE80211BE */
++#ifdef CONFIG_APUP
++	} else if (os_strcmp(buf, "apup") == 0) {
++		bss->apup = !!atoi(pos);
++		if (bss->apup)
++			bss->wds_sta = 1;
++	} else if (os_strcmp(buf, "apup_peer_ifname_prefix") == 0) {
++		os_strlcpy( bss->apup_peer_ifname_prefix,
```
Comment (mwarning) [inline package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch:67]: Please also remove the white space after `os_strlcpy(` and and then end of the call. There are two more other instances I see in this MR.

Comment (G10h4ck) [reply package/network/services/hostapd/patches/780-Implement-APuP-Access-Point-Micro-Peering.patch]: done

**Comment by mwarning** [comment]
Looks good to me - LGTM

**Comment by mwarning** [comment]
@hauke what is the process for merging now?

**Comment by G10h4ck** [comment]
@hauke if you think it is ok please don't let it rot here, re-basing is time consuming.

**Comment by G10h4ck** [comment]
🎉

**Comment by LuisMitaHL** [comment]
Hello, is there any documentation to set up a sample network?

I tried activating “option apup 1” on an Access Point interface on two routers on the same channel, but I could not get it to work.

**Comment by G10h4ck** [comment]
> Hello, is there any documentation to set up a sample network?
> 
> I tried activating “option apup 1” on an Access Point interface on two routers on the same channel, but I could not get it to work.

Do they have the same SSID? Also you need the netowrk to have open authentication, other authentication mechanisms are not supported yet.

**Comment by kofec** [comment]
Here you have example: https://eko.one.pl/forum/viewtopic.php?pid=304802#p304802
You can use translator: https://eko-one-pl.translate.goog/forum/viewtopic.php?pid=304802&_x_tr_sl=pl&_x_tr_tl=en&_x_tr_hl=pl&_x_tr_pto=wapp#p304802

**Comment by G10h4ck** [comment]
They report using it with WPA PSK2, I am surprised it works, I thought it wont work out of the box so i didn't even test it. That is a good news!

**Comment by Sybil127** [comment]
Should the changes to these files really be included as part of the patches?
- src/ap/ubus.h
- src/ap/ubus.c
- src/ap/ucode.h
- src/ap/ucode.c

Considering they originate from the openwrt hostapd package and are not a part of the upstream hostapd source code.

**Comment by BrainSlayer** [comment]
this patch introduces a crashbug. since the interface name in set_wds_sta might now be null (this happens if a wds station disconnects). so every time a wds station disconnection from ap interface hostapd will crash now. 

fix for that issue can be found here https://github.com/mirror/dd-wrt/commit/0c3001a69e8d8300569e416de856c96e903ad130

**Comment by G10h4ck** [comment]
Thanks @BrainSlayer for reporting and for providing a patch, this PR includes your suggestion https://github.com/openwrt/openwrt/pull/16298

**Comment by BrainSlayer** [comment]
@G10h4ck i have seen it. thanks for your efforts
